### ISSUE TITLE
[BUG] Random initializer produces different values inside/outside a gradient tape.

### ISSUE THREAD

**Issue by shishaochen** [tensorflow/tensorflow#54361]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary):  Binary
- TensorFlow version (use command below): 2.8
- Python version: 3.8

**Describe the current behavior**

As the script below shows, no matter API version of random intializer is v1 or v2, values produced are different between inside a gradient tape or not. 

```python
import numpy as np
import tensorflow as tf

w1_init = tf.compat.v1.random_normal_initializer(mean=0., stddev=1., seed=3)
w1_val = w1_init(shape=[4,5], dtype=tf.float64)

w2_init = tf.initializers.RandomNormal(mean=0., stddev=1., seed=3)
w2_val = w2_init(shape=[4,5], dtype=tf.float64)

# assert np.allclose(w1_val, w2_val), 'W1 == W2?'

with tf.GradientTape() as tape:
    w3_init = tf.compat.v1.random_normal_initializer(mean=0., stddev=1., seed=3)
    w3_val = w3_init(shape=[4,5], dtype=tf.float64)
    w3_plus_one = w3_val + 1.

# assert np.allclose(w1_val, w3_val), 'W1 == W3?'

with tf.GradientTape() as tape:
    w4_init = tf.initializers.RandomNormal(mean=0., stddev=1., seed=3)
    w4_val = w4_init(shape=[4,5], dtype=tf.float64)
    w4_plus_one = w4_val + 1.

# assert np.allclose(w2_val, w4_val), 'W2 == W4?'
```

**Describe the expected behavior**

We want `w1_val == w3_val` and `w2_val == w4_val`, then  reproducity can be ensured.

Due to API version change, `w1_val != w2_val` can be accepted.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no

**Comment by shishaochen** [comment]
> @shishaochen , While executing the given code i was not able to found any results which was mentioned in the issue.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/4bf4fc134b8c60bdb01f7e72e31fe791/untitled222.ipynb) and let us know if we are missing anything

@tilakrayal You need uncomment the last 2 assertions and the value inconsistency will occur.

**Comment by jonas-eschle** [comment]
I can confirm this bug and be a bit more precise: the random initializer is not deterministic. Maybe this needs to be switched on?

Here is a snipped, simply using the exact same random initializer (same input arguments and seed) yields completely different results.

https://colab.research.google.com/drive/1tcX4WbsQgnv13OI2krdOPTY7t-1_ZVSa?usp=sharing

The code in plain text
```
import tensorflow as tf

import numpy as np
import tensorflow as tf

dtype = tf.float64  # also for 32
mean = 0.
stddev = 1.
seed = 3
shape = [4,5]

w1_init = tf.initializers.RandomNormal(mean=mean, stddev=stddev, seed=seed)
w1_val = w1_init(shape=shape, dtype=dtype)

w2_init = tf.initializers.RandomNormal(mean=mean, stddev=stddev, seed=seed)
w2_val = w2_init(shape=shape, dtype=dtype)

# print(w1_val, w2_val)
print('Difference', np.abs(w1_val - w2_val))
assert np.allclose(w1_val, w2_val), 'W1 == W2?'

```

**Comment by shishaochen** [comment]
> I can confirm this bug and be a bit more precise: the random initializer is not deterministic (@shishaochen maybe you want to adjust the title accordingly?). Maybe this needs to be switched on?
> 
> Here is a snipped, simply using the exact same random initializer (same input arguments and seed) yields completely different results.
> 
> https://colab.research.google.com/drive/1tcX4WbsQgnv13OI2krdOPTY7t-1_ZVSa?usp=sharing
> 
> The code in plain text
> 
> ```
> import tensorflow as tf
> 
> import numpy as np
> import tensorflow as tf
> 
> dtype = tf.float64  # also for 32
> mean = 0.
> stddev = 1.
> seed = 3
> shape = [4,5]
> 
> w1_init = tf.initializers.RandomNormal(mean=mean, stddev=stddev, seed=seed)
> w1_val = w1_init(shape=shape, dtype=dtype)
> 
> w2_init = tf.initializers.RandomNormal(mean=mean, stddev=stddev, seed=seed)
> w2_val = w2_init(shape=shape, dtype=dtype)
> 
> # print(w1_val, w2_val)
> print('Difference', np.abs(w1_val - w2_val))
> assert np.allclose(w1_val, w2_val), 'W1 == W2?'
> ```

@jonas-eschle There is no error when I execute your codes. My mentioned difference occurs between inisde GradientTape case and outside one.

**Comment by jonas-eschle** [comment]
> @jonas-eschle There is no error when I execute your codes. My mentioned difference occurs between inisde GradientTape case and outside one.

Hm, that is weird, it does fail for me. I understand that you meant the case inside GradientTape, but I suspect to have found (weirdly) a more general bug. Did you check the [colab notebook I've linked](https://colab.research.google.com/drive/1tcX4WbsQgnv13OI2krdOPTY7t-1_ZVSa?usp=sharing)? It fails there, do you agree? So if it doesn't fail for you, which version are you using? Or am I missing something somewhere in my notebook snippet?

**Comment by tilakrayal** [comment]
@chunduriv ,
I was able to reproduce the issue in tf v2.7, v2.8 and nightly.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/f9f93283accd1e1d65c07b3e87fd2b8c/54361.ipynb).

**Comment by shishaochen** [comment]
> > @jonas-eschle There is no error when I execute your codes. My mentioned difference occurs between inisde GradientTape case and outside one.
> 
> Hm, that is weird, it does fail for me. I understand that you meant the case inside GradientTape, but I suspect to have found (weirdly) a more general bug. Did you check the [colab notebook I've linked](https://colab.research.google.com/drive/1tcX4WbsQgnv13OI2krdOPTY7t-1_ZVSa?usp=sharing)? It fails there, do you agree? So if it doesn't fail for you, which version are you using? Or am I missing something somewhere in my notebook snippet?

Your Tensorflow release of version `2.9.0-dev20220214` seems a private one?
My version is the official one `2.8.0`.

**Comment by jonas-eschle** [comment]
> @jonas-eschle Your Tensorflow release of version `2.9.0-dev20220214` seems a private one? My version is the official one `2.8.0`.

I've also tried with the 2.8, but the one 2.9.0-dev20220214 is the "nightly": that's the develop branch. Every night they create a package using the very newest version.

Sure, first thing is to check if it works with the official 2.8. But since that failed, it's good to also test against the nightlies, maybe the bug was just recently fixed.

btw, you see the version number is 2.9.something. So this is "the next one". dev indicates that it is the development version.

To install the nightlies, you can use `pip install tf-nightly` (just for testing purpose, you should not use this normally)

**Comment by wangpengmit** [comment]
@qlzh727

**Comment by wangpengmit** [comment]
This is working-as-intended. Both V1 and V2 (since 2.7) initializers are stateful objects. Calling the same initializer the second time will give different results.

Note that V1 initializers are deprecated and have many problems. Please use V2 initializers.

**Comment by shishaochen** [comment]
> This is working-as-intended. Both V1 and V2 (since 2.7) initializers are stateful objects. Calling the same initializer the second time will give different results.
> 
> Note that V1 initializers are deprecated and have many problems. Please use V2 initializers.

@wangpengmit If the initializer is stateful, the parameter `seed` seems useless？

**Comment by wangpengmit** [comment]
> If the initializer is stateful, the parameter seed seems useless？

No. `seed` determines the random-number *sequence* you'll get from an initializer. Two initializers sharing the same seed will give you the same sequence. (And different seeds lead to different sequences.)

**Comment by shishaochen** [comment]
> > If the initializer is stateful, the parameter seed seems useless？
> 
> No. `seed` determines the random-number _sequence_ you'll get from an initializer. Two initializers sharing the same seed will give you the same sequence. (And different seeds lead to different sequences.)

@wangpengmit You saied "Calling the same initializer the second time will give different results". However, my code snippet created 2 independent initializers with the same seed instead of 1 initializer called twice.

**Comment by wangpengmit** [comment]
You mean the failure of the `assert np.allclose(w2_val, w4_val), 'W2 == W4?'` assertion? Ah sorry I missed that part. It hits the same problem as https://github.com/keras-team/keras/issues/15586#issuecomment-966577975 . Basically a recent change of V2 initializers in TF2.7 to use `tf.random.uniform` etc caused a reproducibility regression (because `tf.random.uniform`'s weird kernel-reused behaviors). It will be fixed in a future release by changing V2 initializers to use `tf.random.Generator` (or back to use `tf.random.stateless_*`). (@qlzh727)

**Comment by shishaochen** [comment]
> You mean the failure of the `assert np.allclose(w2_val, w4_val), 'W2 == W4?'` assertion? Ah sorry I missed that part. It hits the same problem as [keras-team/keras#15586 (comment)](https://github.com/keras-team/keras/issues/15586#issuecomment-966577975) . Basically a recent change of V2 initializers in TF2.7 to use `tf.random.uniform` etc caused a reproducibility regression (because `tf.random.uniform`'s weird kernel-reused behaviors). It will be fixed in a future release by changing V2 initializers to use `tf.random.Generator` (or back to use `tf.random.stateless_*`). (@qlzh727)

Got it!
